### PR TITLE
Plugin for fetching pages using a headless browser

### PR DESCRIPTION
**Pull request by smadha**
- This is a very first version of plugin
- We are looking to improve further by 
 - restricting resources (like css, fonts) as per url filter
 -  Adding resources already downloaded in crawled data. For example while loading a web page on browser images on that web page are also downloaded. We can send these downloaded images to crawldb

### CODE REVIEW

**Comment by karanjeets** [comment]
@smadha Thanks a lot!! Appreciate your hard work  :)
There was a dependency conflict when running fetcher on top of Apache Spark cluster. It is now fixed.

@thammegowda Everything looks good from my end. Tested it locally & on Spark cluster. Do you want to give it a roll before it's merged?

File: sparkler-app/src/main/scala/edu/usc/irds/sparkler/pipeline/FetchFunction.scala [hunk 1]
```diff
@@ -17,55 +17,51 @@
...
     val fetchedAt = new Date()
     val metadata = new Metadata()
     try {
-      val urlConn = new URL(resource.url).openConnection()
-      urlConn.setConnectTimeout(FETCH_TIMEOUT)
-
-      val responseCode = urlConn.asInstanceOf[HttpURLConnection].getResponseCode
-      LOG.debug("STATUS CODE : " + responseCode + " " + resource.url)
+      val fetcher:scala.Option[Fetcher] = PluginService.getExtension(classOf[Fetcher], job)
+      var fetchedData: FetchedData = fetcher.get.fetch(resource.url)
```
Comment (thammegowda) [inline sparkler-app/src/main/scala/edu/usc/irds/sparkler/pipeline/FetchFunction.scala]: ideally, we shouldn't call `Option.get`. Instead we should use `match` pattern and tell how to deal with `null/None`

See how Rahul has correctly handled this [here]( https://github.com/USCDataScience/sparkler/blob/8ccf8088d5fcccc7a8dc5a466553da8753cfbad5/sparkler-app/src/main/scala/edu/usc/irds/sparkler/pipeline/Crawler.scala#L168-L170)

In our case when fetcher is None, we shall fall back to DefaultFetcher

File: sparkler-app/src/main/scala/edu/usc/irds/sparkler/pipeline/FetchFunction.scala [hunk 2]
```diff
@@ -17,55 +17,51 @@
...
+      val rawData: Array[Byte] = fetchedData.getContent
       val status: ResourceStatus = FETCHED
-      val contentType = urlConn.getContentType
+      val contentType = fetchedData.getContentType
       new Content(resource.url, rawData, contentType, rawData.length, Array(),
         fetchedAt, status, metadata)
     } catch {
       case e: Exception =>
         LOG.warn("FETCH-ERROR {}", resource.url)
+        e.printStackTrace()
```
Comment (thammegowda) [inline sparkler-app/src/main/scala/edu/usc/irds/sparkler/pipeline/FetchFunction.scala]: redundant print over Logger.debug

File: sparkler-app/src/main/scala/edu/usc/irds/sparkler/service/PluginService.scala [hunk 3]
```diff
@@ -220,15 +220,21 @@ class PluginService(job:SparklerJob) {
...
-      val service = new PluginService(job)
-      service.load()
-      cache.put(job, service)
+    //lazy initialization for distributed mode (wherever this code gets executed)
+    if (!cache.contains(job.id)){
+      this.synchronized {
+        if(!cache.contains(job.id)) {
+          val service = new PluginService(job)
+          service.load()
+          cache.put(job.id, service)
```
Comment (thammegowda) [inline sparkler-app/src/main/scala/edu/usc/irds/sparkler/service/PluginService.scala]: well done!

File: sparkler-api/src/main/java/edu/usc/irds/sparkler/Fetcher.java [hunk 4]
```diff
@@ -17,10 +17,13 @@
...
 
+import edu.usc.irds.sparkler.model.FetchedData;
+
 /**
  * Created by thammegr on 6/21/16.
  */
 public interface Fetcher extends ExtensionPoint {
 
-    //TODO: this is incomplete
+	FetchedData fetch(String url);
```
Comment (thammegowda) [inline sparkler-api/src/main/java/edu/usc/irds/sparkler/Fetcher.java]: How about:

```
Iterator<FetchedData> fetch(Iterator<String> urls);
```
The idea is that, fetcher may accept stream of URLS, return stream of contents one after the other. It insert fetch delay in between based on the configuration.

I see this is efficient and convenient.

Efficient:  create a single browser/driver/httpclient object and reuse for all fetches one after the other in the stream.
Convenient: When authentication is required, the fetcher can login once, and fetch all the pages in the same host one after the other and logout once it is done(assumption:the urls are grouped by host) . That way we dont get blocked due to too many authentication requests.

**Comment by thammegowda** [comment]
Build failed
```
[INFO] ------------------------------------------------------------------------
[INFO] Building fetcher-jbrowser 0.1-SNAPSHOT
[INFO] ------------------------------------------------------------------------
...... ((Message trimmed))...
[INFO]
[INFO] --- maven-clean-plugin:2.5:clean (default-clean) @ fetcher-jbrowser ---
[INFO]
[INFO] --- maven-bundle-plugin:2.5.0:manifest (default-cli) @ fetcher-jbrowser ---
[WARNING] Manifest edu.usc.irds.sparkler.plugin:fetcher-jbrowser:bundle:0.1-SNAPSHOT : Unused Private-Package instructions, no such package(s) on the class path: [!*]
[ERROR] Manifest edu.usc.irds.sparkler.plugin:fetcher-jbrowser:bundle:0.1-SNAPSHOT : Bundle-Activator not found on the bundle class path nor in imports: edu.usc.irds.sparkler.plugin.FetcherJBrowserActivator
[ERROR] Error(s) found in manifest configuration
[INFO] ------------------------------------------------------------------------
[INFO] Reactor Summary:
[INFO]
[INFO] sparkler-parent .................................... SUCCESS [  0.377 s]
[INFO] sparkler-api ....................................... SUCCESS [  4.851 s]
[INFO] sparkler ........................................... SUCCESS [ 45.358 s]
[INFO] sparkler-plugins ................................... SUCCESS [  0.069 s]
[INFO] urlfilter-regex .................................... SUCCESS [  2.329 s]
[INFO] fetcher-jbrowser ................................... FAILURE [  3.998 s]
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time: 01:00 min
[INFO] Finished at: 2016-10-19T14:46:13-07:00
[INFO] Final Memory: 64M/1499M
[INFO] ------------------------------------------------------------------------
[ERROR] Failed to execute goal org.apache.felix:maven-bundle-plugin:2.5.0:manifest (default-cli) on project fetcher-jbrowser: Error(s) found in manifest configuration -> [Help 1]
[ERROR]
[ERROR] To see the full stack trace of the errors, re-run Maven with the -e switch.
[ERROR] Re-run Maven using the -X switch to enable full debug logging.
[ERROR]
[ERROR] For more information about the errors and possible solutions, please read the following articles:
[ERROR] [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/MojoExecutionException
[ERROR]
[ERROR] After correcting the problems, you can resume the build with the command
[ERROR]   mvn <goals> -rf :fetcher-jbrowser
```

**Comment by karanjeets** [comment]
@thammegowda Can you please try building Sparkler with the following:

```
mvn clean install
```

**Comment by thammegowda** [comment]
@karanjeets  If i do `mvn clean install` 

and then try to run 

```
14:59 $ bin/sparkler.sh inject -su http://www.isjavascriptenabled.com/
>>jobId = sjob-1476914418319
15:00 $ bin/sparkler.sh crawl -id sjob-1476914418319
```

I get 

```
16/10/19 15:01:37 INFO FetchFunction$: FETCHING http://www.isjavascriptenabled.com/
16/10/19 15:01:37 INFO PluginService$: Felix Configuration loaded successfully
Bundle Found: org.apache.felix.framework
16/10/19 15:01:37 WARN FetchFunction$: FETCH-ERROR http://www.isjavascriptenabled.com/
java.util.NoSuchElementException: None.get
	at scala.None$.get(Option.scala:347)
	at scala.None$.get(Option.scala:345)
	at edu.usc.irds.sparkler.pipeline.FetchFunction$.apply(FetchFunction.scala:46)
	at edu.usc.irds.sparkler.pipeline.FetchFunction$.apply(FetchFunction.scala:34)
	at edu.usc.irds.sparkler.pipeline.FairFetcher.next(FairFetcher.scala:52)
	at edu.usc.irds.sparkler.pipeline.FairFetcher.next(FairFetcher.scala:29)
	at scala.collection.Iterator$$anon$12.next(Iterator.scala:444)
	at org.apache.spark.storage.MemoryStore.unrollSafely(MemoryStore.scala:285)
	at org.apache.spark.CacheManager.putInBlockManager(CacheManager.scala:171)
	at org.apache.spark.CacheManager.getOrCompute(CacheManager.scala:78)
	at org.apache.spark.rdd.RDD.iterator(RDD.scala:268)
	at org.apache.spark.rdd.MapPartitionsRDD.compute(MapPartitionsRDD.scala:38)
	at org.apache.spark.rdd.RDD.computeOrReadCheckpoint(RDD.scala:306)
	at org.apache.spark.rdd.RDD.iterator(RDD.scala:270)
	at org.apache.spark.scheduler.ResultTask.runTask(ResultTask.scala:66)
	at org.apache.spark.scheduler.Task.run(Task.scala:89)
	at org.apache.spark.executor.Executor$TaskRunner.run(Executor.scala:214)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at java.lang.Thread.run(Thread.java:745)
16/10/19 15:01:37 INFO ParseFunction$: PARSING  http://www.isjavascriptenabled.com/
```

File: sparkler-app/src/main/scala/edu/usc/irds/sparkler/pipeline/FetchFunction.scala (hunk 1 above)
Comment (karanjeets) [inline sparkler-app/src/main/scala/edu/usc/irds/sparkler/pipeline/FetchFunction.scala]: @thammegowda 👍 

I will work on this.

**Comment by karanjeets** [comment]
@thammegowda This is strange. It worked for me just fine. Let me try with ```bin/sparkler.sh``` and see where things are wrong.

**Comment by karanjeets** [comment]
@thammegowda The issue is with the ```bin/sparkler.sh```

It is using the ```conf``` directory in the build path. As per the new design in Plugin system, the path to bundles directory is generated at compile time using maven resources plugin. See [here](https://github.com/USCDataScience/sparkler/pull/37/files#diff-5af38dcbeba9ee44ecb0a9089219b4cbR68)

Now, as I think, did you add ```conf``` directory in build path to pick post build changes? If yes, I can do something similar with the compiled conf directory.

**Comment by thammegowda** [comment]
@karanjeets 
I am prepending 'con' to priority for configurations. Is there a way to restore that functionality?

**Comment by karanjeets** [comment]
@thammegowda Changes have been made. Please review and merge.

**Comment by thammegowda** [comment]
@karanjeets @smadha merged 💯 . This is a fantastic PR 👍

**Pull request merged**